src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ConstructorParametersShouldMatchPropertyAndFieldNames.cs

```diff
@@ -0,0 +1,292 @@
+                                                                             description: s_localizableDescriptionPropertyPublic,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+        internal static DiagnosticDescriptor FieldPublicRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessageFieldPublic,
+                                                                             DiagnosticCategory.Design,
+                                                                             RuleLevel.BuildWarning,
+                                                                             description: s_localizableDescriptionFieldPublic,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+        internal static DiagnosticDescriptor UnreferencedParameterRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessageUnreferencedParameter,
+                                                                             DiagnosticCategory.Design,
+                                                                             RuleLevel.BuildWarning,
+                                                                             description: s_localizableDescriptionUnreferencedParameter,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(PropertyNameRule, FieldRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction((compilationStartContext) =>
+            {
+                INamedTypeSymbol? jsonConstructorAttributeNamedSymbol = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemTextJsonSerializationJsonConstructorAttribute);
+                if (jsonConstructorAttributeNamedSymbol == null)
+                {
+                    return;
+                }
+
+                var paramAnalyzer = new ParameterAnalyzer(jsonConstructorAttributeNamedSymbol);
```

Comment:
Is the initialization is necessary? seems all references/methods of the this type can be static

Reply:
It's not necessary, but it may improve the readability. Because without it we have to pass json constructor attribute to two methods:

```cs
public static bool ShouldAnalyzeMethod(IMethodSymbol method, INamedTypeSymbol jsonConstructorAttribute)
{
    // We only care about constructors with parameters.
    if (method.Parameters.IsEmpty)
    {
        return false;
    }

    // We only care about constructors that are marked with JsonConstructor attribute.
    return ParameterAnalyzer.IsJsonConstructor(method, jsonConstructorAttribute);
}
```

It means developers should figure out where to get `jsonConstructorAttribute` if they have to deal with those methods. And it may not be obvious why those methods require it. Where providing `jsonConstructorAttribute` during the class initialization follows IoC principle (constructor injection) and also simplifies both methods. IMHO. And it doesn't seem to have a performance impact here, but I may miss something.

Anyway, if you still insist, I've prepared a commit with those changes and can apply it at any time.

Reply:
https://github.com/psxvoid/roslyn-analyzers/commit/97e9017dbc8e1b815cf7a6ba297dac4943be2311